### PR TITLE
Fix incorrect results when writing deletion vectors in Delta Lake

### LINKED ISSUE #23229: Flaky TestDeltaLakeDeleteCompatibility.testDeletionVectors
https://github.com/trinodb/trino/actions/runs/10654242163/job/29530504757

```
tests               | 2024-09-01 19:52:55 WARNING: not retrying; stacktrace does not match pattern '\Q[Databricks][\E(DatabricksJDBCDriver|JDBCDriver)\Q](500593) Communication link failure. Failed to connect to server. Reason: \E((HTTP retry after response received with no Retry-After header, error: HTTP Response code: 503|HTTP Response code: 504), Error message: Unknown.|java.net.SocketTimeoutException: Read timed out.)': [java.lang.AssertionError: Expected row count to be <2>, but was <3>; rows=[[1, 11], [30, -1], [2, -1]]
tests               | 	at io.trino.tests.product.deltalake.TestDeltaLakeDeleteCompatibility.testDeletionVectors(TestDeltaLakeDeleteCompatibility.java:265)
tests               | 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
tests               | 	at java.base/java.lang.Thread.run(Thread.java:1570)
tests               | ]
tests               | 2024-09-01 19:52:55 INFO: Test io.trino.tests.product.deltalake.TestDeltaLakeDeleteCompatibility.testDeletionVectors [none] took 49.84s
tests               | 2024-09-01 19:52:55 INFO: FAILURE     /    io.trino.tests.product.deltalake.TestDeltaLakeDeleteCompatibility.testDeletionVectors [none] (Groups: profile_specific_tests, delta-lake-exclude-91, delta-lake-databricks, delta-lake-oss) took 49.9 seconds
tests               | 2024-09-01 19:52:55 SEVERE: Failure cause:
tests               | java.lang.AssertionError: Expected row count to be <2>, but was <3>; rows=[[1, 11], [30, -1], [2, -1]]
tests               | 	at io.trino.tests.product.deltalake.TestDeltaLakeDeleteCompatibility.testDeletionVectors(TestDeltaLakeDeleteCompatibility.java:265)
tests               | 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
tests               | 	at java.base/java.lang.Thread.run(Thread.java:1570)
```

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes #23229 

## Release notes

```markdown
# Delta Lake
* Fix incorrect results when writing deletion vectors. ({issue}`23229`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 1]
```diff
@@ -365,7 +365,7 @@ private Slice writeMergeResult(Slice path, FileDeletion deletion)
             ParquetMetadata parquetMetadata = MetadataReader.readFooter(dataSource, Optional.empty());
             long rowCount = parquetMetadata.getBlocks().stream().map(BlockMetadata::rowCount).mapToLong(Long::longValue).sum();
             RoaringBitmapArray rowsRetained = new RoaringBitmapArray();
-            rowsRetained.addRange(0, rowCount);
+            rowsRetained.addRange(0, rowCount - 1);
```
Comment (raunaqmorarka) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:368]: Please add javadoc comment to `io.trino.plugin.deltalake.delete.RoaringBitmapArray#addRange` that `rangeEnd` is inclusive

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java [hunk 2]
```diff
@@ -208,7 +208,7 @@ public void testTrinoDeletionVectors()
     }
 
     // Databricks 12.1 and OSS Delta 2.4.0 added support for deletion vectors
-    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS}, dataProvider = "columnMappingModeDataProvider")
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS}, dataProvider = "columnMappingModeDataProvider", invocationCount = 10)
```
Comment (raunaqmorarka) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java:211]: Can we add a test that fails for this case deterministically ?
I don't get why this failure was flaky

**Review by raunaqmorarka — APPROVED** [review]

**Comment by ebyhr** [comment]
The test is still flaky. I'm debugging the issue.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java [hunk 3]
```diff
@@ -267,7 +267,7 @@ public void testDeletionVectors(String mode)
                     .containsOnly(row(2, -1), row(30, -1));
         }
         finally {
-            dropDeltaTableWithRetry("default." + tableName);
+            System.out.println(tableName);
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java:270]: pls drop debug content

Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java:270]: Dropping is intentionally disabled to debug the contents.

**Comment by ebyhr** [comment]
Pushed a new commit to fix correctness issue when rewriting all rows.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 1 above)
Comment (raunaqmorarka) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:368]: Do we have a test which detects this issue ?

**Review by raunaqmorarka — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:368]: The following line throws an exception without this change:
```java
        assertThat(getEntriesFromJson(3, tableLocation + "/_delta_log", FILE_SYSTEM).orElseThrow().get(1).getRemove().deletionVector().orElseThrow())
                .isEqualTo(deletionVector);
```
(Not ideal test, but I assume it's enough)

Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:368]: What was happening as a consequence of having a  too long range before?

**Comment by raunaqmorarka** [comment]
@ebyhr are tables which received writes using this feature in previous releases broken ?
We should add a note in previous releases notes about this and steps to fix if needed.
cc: @mosabua

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java [hunk 4]
```diff
@@ -426,7 +426,8 @@ private Slice writeDeletionVector(
     private Slice onlySourceFile(String sourcePath, FileDeletion deletion)
     {
         String sourceRelativePath = relativePath(rootTableLocation.toString(), sourcePath);
-        DeltaLakeMergeResult result = new DeltaLakeMergeResult(deletion.partitionValues(), Optional.of(sourceRelativePath), Optional.empty());
+        DeletionVectorEntry deletionVector = deletionVectors.get(sourceRelativePath);
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java:429]: follow-up: do we need any special handling for deletion vectors from shallowly cloned tables?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/RemoveFileEntry.java [hunk 5]
```diff
@@ -16,17 +16,20 @@
...
 
 import static java.util.Objects.requireNonNull;
 
 public record RemoveFileEntry(
         String path,
         @Nullable Map<String, String> partitionValues,
         long deletionTimestamp,
-        boolean dataChange)
+        boolean dataChange,
+        Optional<DeletionVectorEntry> deletionVector)
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/RemoveFileEntry.java:28]: Where do we need `deletionVector` in the `remove` file entries?

Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/RemoveFileEntry.java:28]: Just noticed

```
        Optional<DeletionVectorEntry> deletionVector = Optional.empty();
        if (deletionVectorsEnabled) {
            deletionVector = Optional.ofNullable(remove.getRow("deletionVector"))
                    .map(row -> parseDeletionVectorFromParquet(session, row, removeDeletionVectorType.orElseThrow()));
        }
```

**Pull request merged**